### PR TITLE
Move more tests to use the test lib

### PR DESCRIPTION
**Pull request by michaelpj**
- A couple of generalizations to the testlib.
- Move the tests in `core-to-plc` and `plutus-th` to use the testlib.

I think possibly the testlib should go in its own package...

### CODE REVIEW

File: language-plutus-core/common/Common.hs [hunk 1]
```diff
@@ -19,32 +19,32 @@ import           System.FilePath            ((</>))
...
-testNestedGolden folderName =
+-- | Descend into a folder of tests.
+testNested :: String -> [TestNested] -> TestNested
+testNested folderName =
     Reader.local (++ [folderName]) . fmap (testGroup folderName) . sequence
 
 -- | Check the contents of a file against a 'Text'.
-goldenVsText :: TestName -> FilePath -> Text -> TestTree
-goldenVsText name ref = goldenVsString name ref . pure . BSL.fromStrict . encodeUtf8
+goldenVsText :: TestName -> FilePath -> IO Text -> TestTree
```
Comment (effectfully) [inline language-plutus-core/common/Common.hs]: I'd prefer if we keep `goldenVsText ` that receive sa `Text` and define `goldenVsTextM` that receives a `IO Text`. The pure case is much more common and `goldenVsText name ref text` is a much nicer call than `goldenVsText name ref (pure text)`.

File: language-plutus-core/common/Common.hs [hunk 2]
```diff
@@ -19,32 +19,32 @@ import           System.FilePath            ((</>))
 import           Test.Tasty
 import           Test.Tasty.Golden
 
--- | A 'TestTree' of golden tests stored in some folder.
-type TestNestedGolden = Reader [String] TestTree
+-- | A 'TestTree' of tests stored in some folder.
+type TestNested = Reader [String] TestTree
```
Comment (effectfully) [inline language-plutus-core/common/Common.hs]: Are there any tests stored in a folder that are not golden? The `[String]` part specifies just a sequence of names, so we indeed can generically call this `TestNested`, but then I guess we should drop file system terms as well.

File: plutus-th/test/Spec.hs [hunk 3]
```diff
@@ -7,30 +7,24 @@
...
-golden name value = (goldenVsString name ("test/" ++ name ++ ".plc.golden") . pure . BSL.fromStrict . encodeUtf8 . PLC.docText . PLC.prettyPlcClassicDebug . getAst) value
+main = defaultMain $ runTestNestedIn ["test"] tests
 
+golden :: String -> PlcCode -> TestNested
+golden name value = (nestedGoldenVsDoc name . pure . PLC.prettyPlcClassicDebug . getAst) value
 
-tests :: TestTree
-tests = testGroup "Plutus TH frontend" [
+tests :: TestNested
+tests = testGroup "plutus-th" <$> sequence [
```
Comment (effectfully) [inline plutus-th/test/Spec.hs]: I guess there is a specific reason why you do not use `testNested` here?

File: language-plutus-core/common/Common.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/common/Common.hs]: Sure - having a nested name makes sense in other cases too.

File: plutus-th/test/Spec.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-th/test/Spec.hs]: I didn't want to introduce a pointless additional folder. We're already in `test`, I don't want another folder just to group all the top-level things.

File: language-plutus-core/common/Common.hs (hunk 1 above)
Comment (michaelpj) [reply language-plutus-core/common/Common.hs]: If you like - but I think exposing the full underlying API is nice. And the changes required by this are in practice very modest.

File: language-plutus-core/common/Common.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/common/Common.hs]: I just noticed that there was not in fact anything golden-specific about these functions, and removing "golden" from all the names removes some noise.

File: language-plutus-core/common/Common.hs (hunk 1 above)
Comment (effectfully) [reply language-plutus-core/common/Common.hs]: Yeah, but the changes do not look like improvements to me and we will have more tests of this kind.

File: language-plutus-core/common/Common.hs (hunk 2 above)
Comment (effectfully) [reply language-plutus-core/common/Common.hs]: I mean that in this comment:

> A 'TestTree' of tests stored in some folder.

which talks about folders while the type itself is generic and the concept of the "folder" can be irrelevant. I like the generalization, just let's adapt comments as well.

File: plutus-th/test/Spec.hs (hunk 3 above)
Comment (effectfully) [reply plutus-th/test/Spec.hs]: I'd very much prefer to keep golden files separately from machineries that use them. Even better: always keep golden files (nested or not) inside a "Golden" folder. Right now we sometimes interleave golden files with Haskell files and this looks like a mess and is not nice to navigate.

Comment (michaelpj) [reply plutus-th/test/Spec.hs]: They are separate - due to the nesting machinery I have `test/a`, `test/b` etc. I just don't see a point in having `test/golden/a`, `test/golden/b`.

File: language-plutus-core/common/Common.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/common/Common.hs]: Done.

File: language-plutus-core/common/Common.hs (hunk 1 above)
Comment (michaelpj) [reply language-plutus-core/common/Common.hs]: Done. I think the result is worse - now we just have lots of extra functions which just compose with `pure` but I don't care that much.

**Comment by michaelpj** [comment]
Looks like hydra is failing to report notifications again. Let me know if you're happy and I'll merge.

**Comment by effectfully** [comment]
LGTM, feel free to merge.

**Pull request merged**